README.md
[![Build Status](https://travis-ci.org/njh/ruby-mqtt.svg)](https://travis-ci.org/ccutrer/ruby-mqtt)

ruby-mqtt
=========

Pure Ruby gem that implements the [MQTT] protocol, a lightweight protocol for publish/subscribe messaging.

Also includes a class for parsing and generating [MQTT-SN] packets.


Table of Contents
-----------------
* [Installation](#installation)
* [Quick Start](#quick-start)
* [Library Overview](#library-overview)
* [Resources](#resources)
* [License](#license)
* [Contact](#contact)


Installation
------------

You may get the latest stable version from [Rubygems]:

    $ gem install mqtt-ccutrer

Quick Start
-----------

~~~ ruby
require 'mqtt'

# Publish example
MQTT::Client.connect('test.mosquitto.org') do |c|
  c.publish('test', 'message')
end

# Subscribe example
MQTT::Client.connect('test.mosquitto.org') do |c|
  # If you pass a block to the get method, then it will loop
  c.subscribe('test')
  c.get do |packet|
    puts "#{packet.topic}: #{packet.payload}"
  end
end
~~~


Library Overview
----------------

### Connecting ###

A new client connection can be created by passing either a [MQTT URI], a host and port or by passing a hash of attributes.

~~~ ruby
client = MQTT::Client.connect('mqtt://myserver.example.com')
client = MQTT::Client.connect('mqtts://user:pass@myserver.example.com')
client = MQTT::Client.connect('myserver.example.com')
client = MQTT::Client.connect('myserver.example.com', 18830)
client = MQTT::Client.connect(host: 'myserver.example.com', port: 1883 ... )
~~~

TLS/SSL is not enabled by default, to enabled it, pass `ssl: true`:

~~~ ruby
client = MQTT::Client.connect(
  host: 'test.mosquitto.org',
  port: 8883,
  ssl: true
)
~~~

Alternatively you can create a new Client object and then configure it by setting attributes. This example shows setting up client certificate based authentication:

~~~ ruby
client = MQTT::Client.new
client.host = 'myserver.example.com'
client.ssl = true
client.cert_file = path_to('client.pem')
client.key_file  = path_to('client.key')
client.ca_file   = path_to('root-ca.pem')
client.connect
~~~

The connection can either be made without the use of a block:

~~~ ruby
client = MQTT::Client.connect('test.mosquitto.org')
# perform operations
client.disconnect
~~~

Or, if using a block, with an implicit disconnection at the end of the block.

~~~ ruby
MQTT::Client.connect('test.mosquitto.org') do |client|
  # perform operations
end
~~~

For more information, see the list of attributes for the [MQTT::Client] class and the [MQTT::Client.connect] method.


### Publishing ###

To send a message to a topic, use the ```publish``` method:

~~~ ruby
client.publish(topic, payload, retain: false, qos: 0)
~~~

The method will return once the message has been sent to the MQTT server for QoS 0,
or once an Ack has been received from the server for QoS 1.

For more information see the [MQTT::Client#publish] method.


### Subscribing ###

You can send a subscription request to the MQTT server using the subscribe method. One or more [Topic Filters] may be passed in:

~~~ ruby
client.subscribe( 'topic1' )
client.subscribe( 'topic1', 'topic2' )
client.subscribe( 'foo/#' )
~~~

For more information see the [MQTT::Client#subscribe] method.


### Receiving Messages ###

To receive a message, use the get method. This method will block until a message is available. You can access details
about the packet such as if it was retained, the topic it was sent to, and the payload.

~~~ ruby
packet = client.get
packet.retained?
packet.topic
packet.payload
~~~

Alternatively, you can give the get method a block, which will be called for every message received and loop forever:

~~~ ruby
client.get do |packet|
  # Block is executed for every message received
end
~~~

For more information see the [MQTT::Client#get] method.


### Parsing and serialising of packets ###

The parsing and serialising of MQTT and MQTT-SN packets is a separate lower-level API.
You can use it to build your own clients and servers, without using any of the rest of the
code in this gem.

~~~ ruby
# Parse a string containing a binary packet into an object
packet_obj = MQTT::Packet.parse(binary_packet)
    
# Write a PUBACK packet to an IO handle
ios << MQTT::Packet::Puback(id: 20)
    
# Write an MQTT-SN Publish packet with QoS -1 to a UDP socket
socket = UDPSocket.new
socket.connect('localhost', MQTT::SN::DEFAULT_PORT)
socket << MQTT::SN::Packet::Publish.new(
  topic_id: 'TT',
  topic_id_type: :short,
  data: "The time is: #{Time.now}",
  qos: -1
)
socket.close
~~~

Limitations
-----------

 * QoS 2 is not currently supported by client
 * Pending publishes are only persisted as long as a reconnect occurs within
   the configured timeout

Resources
---------

* API Documentation: http://rubydoc.info/gems/mqtt
* Protocol Specification v3.1.1: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html
* Protocol Specification v3.1: http://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html
* MQTT-SN Protocol Specification v1.2: http://mqtt.org/new/wp-content/uploads/2009/06/MQTT-SN_spec_v1.2.pdf
* MQTT Homepage: http://www.mqtt.org/
* GitHub Project: http://github.com/njh/ruby-mqtt


License
-------

The mqtt ruby gem is licensed under the terms of the MIT license.
See the file LICENSE for details.


[MQTT]:           http://www.mqtt.org/
[MQTT-SN]:        http://mqtt.org/2013/12/mqtt-for-sensor-networks-mqtt-sn
[Rubygems]:       http://rubygems.org/
[Bundler]:        http://bundler.io/
[MQTT URI]:       https://github.com/mqtt/mqtt.github.io/wiki/URI-Scheme
[Topic Filters]:  http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html#_Toc388534397

[MQTT::Client]:           http://rubydoc.info/gems/mqtt-ccutrer/MQTT/Client#instance_attr_details
[MQTT::Client.connect]:   http://rubydoc.info/gems/mqtt-ccutrer/MQTT/Client.connect
[MQTT::Client#publish]:   http://rubydoc.info/gems/mqtt-ccutrer/MQTT/Client:publish
[MQTT::Client#subscribe]: http://rubydoc.info/gems/mqtt-ccutrer/MQTT/Client:subscribe
[MQTT::Client#get]:       http://rubydoc.info/gems/mqtt-ccutrer/MQTT/Client:get



NEWS.md
Ruby MQTT NEWS
==============

Ruby MQTT Version 1.0.0 (2021-02-19) (CCutrer fork)
---------------------------------------------------
* Increase minimum Ruby version to 2.5
* Change several methods to take kwargs
* Remove `get` vs. `get_packet`; just always return packets from get
* `get` no longer takes an optional topic to subscribe to; just call
  `subscribe`
* Remove `last_ping_response` since it's not recorded anymore
* Remove deprecated `remote_host` and `remote_port`
* `connect` no longer takes an optional client id; just set it beforehand
* Add `batch_publish` method for improving performance of QoS 1 publishes
* Add `flush` method to ensure writes have finished, since writes are
  now done in a background thread
* Add `wait_for_ack` param to `subscribe` and `unsubscribe`
* Automatic reconnects
* Re-send Pub, Sub, and Unsub packets that don't get acked, even across
  server disconnects
* Only send keepalive pings if we haven't received any other packets,
  and measure from the last packet received, instead of the last ping
  sent
* Improve CPU usage by only checking for timeouts at the time of the
  timeout, instead of polling

Ruby MQTT Version 0.5.0 (2017-04-16)
------------------------------------

* Switched default protocol version to 3.1.1
* Added support for Server Name Identification (SNI)
* Fix for unescaping user/password in URI
* Fix for bug in MQTT::Proxy class
* Add the ability to ignore retained packets when subscribed.
* Fix problem of wrong Puback packet ID
* Don't keepalive ping if disconnected
* Immediately close socket after failed Connack
* FakeServer improvements
* Fix for working with mathn library.


Ruby MQTT Version 0.4.0 (2016-06-27)
------------------------------------

* Added puback handling for QoS level 1
* Low-level MQTT-SN packet parsing support
* Allow certs to be set directly instead of just by file
* Allow keyphrase for certs to be passed through
* Put 'disconnect' inside an 'ensure' block
* Fix for error on publish with frozen payload
* Fix for packets always getting id 1
* Improvements to tests


Ruby MQTT Version 0.3.1 (2014-10-10)
------------------------------------

* Added ```last_ping_response``` to attribute to ```MQTT::Client```


Ruby MQTT Version 0.3.0 (2014-08-26)
------------------------------------

* Added support for MQTT protocol version 3.1.1
* Renamed a number of methods/attributes:
  - Renamed ```:granted_qos``` to ```:return_codes```
  - Renamed ```:remote_port``` to ```:port```
  - Renamed ```:remote_host``` to ```:host```
  - Renamed ```:message_id``` to ```:id```
  - Renamed ```:protocol_version``` to ```:protocol_level```
  - Renamed ```MQTT_BROKER``` environment variable to ```MQTT_SERVER```
* Added more checks to ensure that the 3.1.1 protocol specs are adhered to
* Added a Library Overview section to the README
* Added links to the protocol specification to README
* Improvements to the YARD API documentation
* Don't display payload in inspect if it contains non-visible ASCII characters
* Upgraded to rspec 3
* Various minor bug fixes and corrections


Ruby MQTT Version 0.2.0 (2014-04-02)
------------------------------------

* Added SSL/TLS support
* Added support for passing connection details using a URI
* Added support for using the ```MQTT_BROKER``` environment variable
* Allow passing array of topics to Client#unsubscribe
* Allow more combinations of arguments to be passed to a new Client
* No longer defaults to ‘localhost’ if there is no server configured
* Fixed more 'unused variable' warnings
* Documentation improvements
* Ruby 1.8 fixes
* Ruby 2 fixes


Ruby MQTT Version 0.1.0 (2013-09-07)
------------------------------------

* Changed license to MIT, to simplify licensing concerns
* Improvements for UTF-8 handling under Ruby 1.9
* Added ```get_packet``` method
* Added support for a keep-alive value of 0
* Added a #inspect method to the Packet classes
* Added checks for the protocol name and version
* Added check to ensure that packet body isn't too big
* Added validation of QoS value
* Added example of using authentication
* Fixed 'unused variable' warnings
* Reduced duplicated code in packet parsing
* Improved testing
  - Created fake server and integration tests
  - Better test coverage
  - Added more tests for error states


Ruby MQTT Version 0.0.9 (2012-12-21)
------------------------------------

* Fixes for Ruby 1.9.3 by Mike English
* Fix for ```client_id``` typo by Anubisss
* Added methods to inspect the incoming message queue: ```queue_empty?``` and ```queue_length```
* Fixed incorrect implementation of the parsing and serialising of Subscription Acknowledgement packets

lib/mqtt/client.rb
# frozen_string_literal: true

autoload :OpenSSL, 'openssl'
autoload :SecureRandom, 'securerandom'
autoload :URI, 'uri'

# Client class for talking to an MQTT server
module MQTT
  class Client
    # Hostname of the remote server
    attr_accessor :host

    # Port number of the remote server
    attr_accessor :port

    # The version number of the MQTT protocol to use (default 3.1.1)
    attr_accessor :version

    # Set to true to enable SSL/TLS encrypted communication
    #
    # Set to a symbol to use a specific variant of SSL/TLS.
    # Allowed values include:
    #
    # @example Using TLS 1.0
    #    client = Client.new('mqtt.example.com', ssl: :TLSv1)
    # @see OpenSSL::SSL::SSLContext::METHODS
    attr_accessor :ssl

    # Time (in seconds) between pings to remote server (default is 15 seconds)
    attr_accessor :keep_alive

    # Set the 'Clean Session' flag when connecting? (default is true)
    attr_accessor :clean_session

    # Client Identifier
    attr_accessor :client_id

    # Number of seconds to wait for acknowledgement packets (default is 5 seconds)
    attr_accessor :ack_timeout

    # How many times to attempt re-sending packets that weren't acknowledged
    # (default is 5) before giving up
    attr_accessor :resend_limit

    # How many attempts to re-establish a connection after it drops before
    # giving up (default 5)
    attr_accessor :reconnect_limit

    # How long to wait between re-connection attempts (exponential - i.e.
    # immediately after first drop, then 5s, then 25s, then 125s, etc.
    # when theis value defaults to 5)
    attr_accessor :reconnect_backoff

    # Username to authenticate to the server with
    attr_accessor :username

    # Password to authenticate to the server with
    attr_accessor :password

    # The topic that the Will message is published to
    attr_accessor :will_topic

    # Contents of message that is sent by server when client disconnect
    attr_accessor :will_payload

    # The QoS level of the will message sent by the server
    attr_accessor :will_qos

    # If the Will message should be retain by the server after it is sent
    attr_accessor :will_retain

    # Default attribute values
    ATTR_DEFAULTS = {
      host: nil,
      port: nil,
      version: '3.1.1',
      keep_alive: 15,
      clean_session: true,
      client_id: nil,
      ack_timeout: 5,
      resend_limit: 5,
      reconnect_limit: 5,
      reconnect_backoff: 5,
      username: nil,
      password: nil,
      will_topic: nil,
      will_payload: nil,
      will_qos: 0,
      will_retain: false,
      ssl: false
    }.freeze

    # Create and connect a new MQTT Client
    #
    # Accepts the same arguments as creating a new client.
    # If a block is given, then it will be executed before disconnecting again.
    #
    # Example:
    #  MQTT::Client.connect('myserver.example.com') do |client|
    #    # do stuff here
    #  end
    #
    def self.connect(*args, &block)
      client = MQTT::Client.new(*args)
      client.connect(&block)
      client
    end

    # Generate a random client identifier
    # (using the characters 0-9 and a-z)
    def self.generate_client_id(prefix = 'ruby', length = 16)
      "#{prefix}#{SecureRandom.alphanumeric(length).downcase}"
    end

    # Create a new MQTT Client instance
    #
    # Accepts one of the following:
    # - a URI that uses the MQTT scheme
    # - a hostname and port
    # - a Hash containing attributes to be set on the new instance
    #
    # If no arguments are given then the method will look for a URI
    # in the MQTT_SERVER environment variable.
    #
    # Examples:
    #  client = MQTT::Client.new
    #  client = MQTT::Client.new('mqtt://myserver.example.com')
    #  client = MQTT::Client.new('mqtt://user:pass@myserver.example.com')
    #  client = MQTT::Client.new('myserver.example.com')
    #  client = MQTT::Client.new('myserver.example.com', 18830)
    #  client = MQTT::Client.new(host: 'myserver.example.com')
    #  client = MQTT::Client.new(host: 'myserver.example.com', keep_alive: 30)
    #
    def initialize(host = nil, port = nil, **attributes)
      # Set server URI from environment if present
      if host.nil? && port.nil? && attributes.empty? && ENV['MQTT_SERVER']
        attributes.merge!(parse_uri(ENV['MQTT_SERVER']))
      end

      if host
        case host
        when URI, %r{^mqtts?://}
          attributes.merge!(parse_uri(host))
        else
          attributes[:host] = host
        end
      end

      attributes[:port] = port unless port.nil?

      # Merge arguments with default values for attributes
      ATTR_DEFAULTS.merge(attributes).each_pair do |k, v|
        send("#{k}=", v)
      end

      # Set a default port number
      if @port.nil?
        @port = @ssl ? MQTT::DEFAULT_SSL_PORT : MQTT::DEFAULT_PORT
      end

      # Initialise private instance variables
      @socket = nil
      @read_queue = Queue.new
      @write_queue = Queue.new

      @read_thread = nil
      @write_thread = nil

      @acks = {}

      @connection_mutex = Mutex.new
      @acks_mutex = Mutex.new
      @wake_up_pipe = IO.pipe

      @connected = false
    end

    # Get the OpenSSL context, that is used if SSL/TLS is enabled
    def ssl_context
      @ssl_context ||= OpenSSL::SSL::SSLContext.new
    end

    # Set a path to a file containing a PEM-format client certificate
    def cert_file=(path)
      self.cert = File.read(path)
    end

    # PEM-format client certificate
    def cert=(cert)
      ssl_context.cert = OpenSSL::X509::Certificate.new(cert)
    end

    # Set a path to a file containing a PEM-format client private key
    def key_file=(*args)
      path, passphrase = args.flatten
      ssl_context.key = OpenSSL::PKey::RSA.new(File.open(path), passphrase)
    end

    # Set to a PEM-format client private key
    def key=(*args)
      cert, passphrase = args.flatten
      ssl_context.key = OpenSSL::PKey::RSA.new(cert, passphrase)
    end

    # Set a path to a file containing a PEM-format CA certificate and enable peer verification
    def ca_file=(path)
      ssl_context.ca_file = path
      ssl_context.verify_mode = OpenSSL::SSL::VERIFY_PEER unless path.nil?
    end

    # Set the Will for the client
    #
    # The will is a message that will be delivered by the server when the client dies.
    # The Will must be set before establishing a connection to the server
    def set_will(topic, payload, retain: false, qos: 0)
      self.will_topic = topic
      self.will_payload = payload
      self.will_retain = retain
      self.will_qos = qos
    end

    # Connect to the MQTT server
    #
    # If a block is given, then yield to that block and then disconnect again.
    def connect
      if connected?
        yield(self) if block_given?
        return
      end

      if @client_id.nil? || @client_id.empty?
        raise 'Must provide a client_id if clean_session is set to false' unless @clean_session

        # Empty client id is not allowed for version 3.1.0
        @client_id = MQTT::Client.generate_client_id if @version == '3.1.0'
      end

      raise ArgumentError, 'No MQTT server host set when attempting to connect' if @host.nil?

      connect_internal

      return unless block_given?

      # If a block is given, then yield and disconnect
      begin
        yield(self)
      ensure
        disconnect
      end
    end

    # wait until all messages have been sent
    def flush
      raise NotConnectedException unless connected?

      queue = Queue.new
      @write_queue << queue
      queue.pop
      nil
    end

    # Disconnect from the MQTT server.
    #
    # If you don't want to say goodbye to the server, set send_msg to false.
    def disconnect(send_msg: true)
      return unless connected?

      @read_queue << [ConnectionClosedException.new, current_time]
      # Stop reading packets from the socket first
      @connection_mutex.synchronize do
        if @write_thread&.alive?
          @write_thread.kill
          @write_thread.join
        end
        @read_thread.kill if @read_thread&.alive?
        @read_thread = @write_thread = nil

        @connected = false
      end
      @acks_mutex.synchronize do
        @acks.each_value do |pending_ack|
          pending_ack.queue << :close
        end
        @acks.clear
      end

      return unless @socket

      if send_msg
        packet = MQTT::Packet::Disconnect.new
        begin
          @socket.write(packet.to_s)
        rescue
          nil
        end
      end
      @socket.close
      @socket = nil
    end

    # Checks whether the client is connected to the server.
    #
    # Note that this returns true even if the connection is down and we're
    # trying to reconnect
    def connected?
      @connected
    end

    # registers a callback to be called when a connection is re-established
    #
    # can be used to re-subscribe (if you're not using persistent sessions)
    # to topics, and/or re-publish aliveness (if you set a Will)
    def on_reconnect(&block)
      @on_reconnect = block
    end

    # yields a block, and after the block returns all messages are
    # published at once, waiting for any necessary PubAcks for QoS 1
    # packets as a batch at the end
    #
    #  For example:
    #    client.batch_publish do
    #      client.publish("topic1", "value1", qos: 1)
    #      client.publish("topic2", "value2", qos: 1)
    #    end
    def batch_publish
      return yield if @batch_publish

      @batch_publish = {}

      begin
        yield

        batch = @batch_publish
        @batch_publish = nil
        batch.each do |(kwargs, values)|
          publish(values, **kwargs)
        end
      ensure
        @batch_publish = nil
      end
    end

    # Publish a message on a particular topic to the MQTT server.
    def publish(topics, payload = nil, retain: false, qos: 0)
      if topics.is_a?(Hash) && !payload.nil?
        raise ArgumentError, 'Payload cannot be passed if passing a hash for topics and payloads'
      end
      raise NotConnectedException unless connected?

      if @batch_publish && qos != 0
        values = @batch_publish[{ retain: retain, qos: qos }] ||= {}
        if topics.is_a?(Hash)
          values.merge!(topics)
        else
          values[topics] = payload
        end
        return
      end

      pending_acks = []

      topics = { topics => payload } unless topics.is_a?(Hash)

      topics.each do |(topic, topic_payload)|
        raise ArgumentError, 'Topic name cannot be nil' if topic.nil?
        raise ArgumentError, 'Topic name cannot be empty' if topic.empty?

        packet = MQTT::Packet::Publish.new(
          id: next_packet_id,
          qos: qos,
          retain: retain,
          topic: topic,
          payload: topic_payload
        )

        pending_acks << register_for_ack(packet) unless qos.zero?

        # Send the packet
        send_packet(packet)
      end

      return if qos.zero?

      pending_acks.each do |ack|
        wait_for_ack(ack)
      end
      nil
    end

    # Send a subscribe message for one or more topics on the MQTT server.
    # The topics parameter should be one of the following:
    # * String: subscribe to one topic with QoS 0
    # * Array: subscribe to multiple topics with QoS 0
    # * Hash: subscribe to multiple topics where the key is the topic and the value is the QoS level
    #
    # For example:
    #   client.subscribe( 'a/b' )
    #   client.subscribe( 'a/b', 'c/d' )
    #   client.subscribe( ['a/b',0], ['c/d',1] )
    #   client.subscribe( { 'a/b' => 0, 'c/d' => 1 } )
    #
    def subscribe(*topics, wait_for_ack: false)
      raise NotConnectedException unless connected?

      packet = MQTT::Packet::Subscribe.new(
        id: next_packet_id,
        topics: topics
      )
      token = register_for_ack(packet) if wait_for_ack
      send_packet(packet)
      wait_for_ack(token) if wait_for_ack
    end

    # Send a unsubscribe message for one or more topics on the MQTT server
    def unsubscribe(*topics, wait_for_ack: false)
      raise NotConnectedException unless connected?

      topics = topics.first if topics.is_a?(Enumerable) && topics.count == 1

      packet = MQTT::Packet::Unsubscribe.new(
        topics: topics,
        id: next_packet_id
      )
      token = register_for_ack(packet) if wait_for_ack
      send_packet(packet)
      wait_for_ack(token) if wait_for_ack
    end

    # Return the next message received from the MQTT server.
    #
    # The method either returns the Publish packet:
    #   packet = client.get
    #
    # Or can be used with a block to keep processing messages:
    #   client.get do |packet|
    #     # Do stuff here
    #   end
    #
    def get
      raise NotConnectedException unless connected?

      loop_start = current_time
      loop do
        packet = @read_queue.pop
        if packet.is_a?(Array) && packet.last >= loop_start
          e = packet.first
          e.set_backtrace((e.backtrace || []) + ['<from MQTT worker thread>'] + caller)
          raise e
        end
        next unless packet.is_a?(Packet)

        unless block_given?
          puback_packet(packet) if packet.qos > 0
          return packet
        end

        yield packet
        puback_packet(packet) if packet.qos > 0
      end
    end

    # Returns true if the incoming message queue is empty.
    def queue_empty?
      @read_queue.empty?
    end

    # Returns the length of the incoming message queue.
    def queue_length
      @read_queue.length
    end

    # Clear the incoming message queue.
    def clear_queue
      @read_queue.clear
    end

    private

    PendingAck = Struct.new(:packet, :queue, :timeout_at, :send_count)

    def connect_internal
      # Create network socket
      tcp_socket = TCPSocket.new(@host, @port)

      if @ssl
        # Set the protocol version
        ssl_context.ssl_version = @ssl if @ssl.is_a?(Symbol)

        @socket = OpenSSL::SSL::SSLSocket.new(tcp_socket, ssl_context)
        @socket.sync_close = true

        # Set hostname on secure socket for Server Name Indication (SNI)
        @socket.hostname = @host if @socket.respond_to?(:hostname=)

        @socket.connect
      else
        @socket = tcp_socket
      end

      # Construct a connect packet
      packet = MQTT::Packet::Connect.new(
        version: @version,
        clean_session: @clean_session,
        keep_alive: @keep_alive,
        client_id: @client_id,
        username: @username,
        password: @password,
        will_topic: @will_topic,
        will_payload: @will_payload,
        will_qos: @will_qos,
        will_retain: @will_retain
      )

      # Send packet
      @socket.write(packet.to_s)

      # Receive response
      receive_connack

      @connected = true

      @write_thread = Thread.new do
        while (packet = @write_queue.pop)
          # flush command
          if packet.is_a?(Queue)
            packet << :flushed
            next
          end
          @socket.write(packet.to_s)
        end
      rescue => e
        @write_queue << packet if packet
        reconnect(e)
      end

      @read_thread = Thread.new do
        receive_packet while connected?
      end
    end

    def reconnect(exception)
      should_exit = nil
      @connection_mutex.synchronize do
        @socket&.close
        @socket = nil
        @read_thread&.kill if Thread.current != @read_thread
        @write_thread&.kill if Thread.current != @write_thread
        should_exit = Thread.current == @read_thread
        @read_thread = @write_thread = nil

        retries = 0
        begin
          connect_internal unless @reconnect_limit == 0
        rescue
          @socket&.close
          @socket = nil

          if (retries += 1) < @reconnect_limit
            sleep @reconnect_backoff ** retries
            retry
          end
        end

        unless @socket
          # couldn't reconnect
          @acks_mutex.synchronize do
            @acks.each_value do |pending_ack|
              pending_ack.queue << :close
            end
            @acks.clear
          end
          @connected = false
          @read_queue << [exception, current_time]
          return
        end
      end

      begin
        if @on_reconnect&.arity == 0
          @on_reconnect.call
        else
          @on_reconnect&.call(@connack)
        end
      rescue => e
        @read_queue << [e, current_time]
        disconnect
      end
      Thread.exit if should_exit
    end

    # Try to read a packet from the server
    # Also sends keep-alive ping packets.
    def receive_packet
      # Poll socket - is there data waiting?
      timeout = next_timeout
      read_ready, = IO.select([@socket, @wake_up_pipe[0]], [], [], timeout)

      # we just needed to break out of our select to set up a new timeout;
      # we can discard the actual contents
      @wake_up_pipe[0].readpartial(4096) if read_ready&.include?(@wake_up_pipe[0])

      handle_timeouts

      if read_ready&.include?(@socket)
        packet = MQTT::Packet.read(@socket)
        handle_packet(packet)
      end

      handle_keep_alives
    rescue => e
      reconnect(e)
    end

    def register_for_ack(packet)
      queue = Queue.new

      timeout_at = current_time + @ack_timeout
      @acks_mutex.synchronize do
        if @acks.empty?
          # just need to wake up the read thread to set up the timeout for this packet
          @wake_up_pipe[1].write('z')
        end
        @acks[packet.id] = PendingAck.new(packet, queue, timeout_at, 1)
      end
    end

    def wait_for_ack(pending_ack)
      response = pending_ack.queue.pop
      case response
      when :close
        raise ConnectionClosedException
      when :resend_limit_exceeded
        raise ResendLimitExceededException
      end
    end

    def handle_packet(packet)
      @last_packet_received_at = current_time
      @keep_alive_sent = false
      case packet
      when MQTT::Packet::Publish
        # Add to queue
        @read_queue.push(packet)
      when MQTT::Packet::Pingresp
        # do nothing; setting @last_packet_received_at already handled it
      when MQTT::Packet::Puback,
        MQTT::Packet::Suback,
        MQTT::Packet::Unsuback
        @acks_mutex.synchronize do
          pending_ack = @acks[packet.id]
          if pending_ack
            @acks.delete(packet.id)
            pending_ack.queue << packet
          end
        end
      end
      # Ignore all other packets
      # FIXME: implement responses for QoS  2
    end

    def handle_timeouts
      @acks_mutex.synchronize do
        current_time = self.current_time
        @acks.each_value do |pending_ack|
          break unless pending_ack.timeout_at <= current_time

          resend(pending_ack)
        end
      end
    end

    def resend(pending_ack)
      packet = pending_ack.packet
      if (pending_ack.send_count += 1) > @resend_limit
        @acks.delete(packet.id)
        pending_ack.queue << :resend_limit_exceeded
        return
      end
      # timed out, or simple re-send
      @wake_up_pipe[1].write('z') if @acks.first.first == packet.id
      pending_ack.timeout_at = current_time + @ack_timeout
      packet.duplicate = true
      send_packet(packet)
    end

    def current_time
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end

    def next_timeout
      timeout_from_acks = @acks_mutex.synchronize do
        @acks.first&.last&.timeout_at
      end
      return nil if timeout_from_acks.nil? && @keep_alive.nil?

      next_ping = @last_packet_received_at + @keep_alive if @keep_alive && !@keep_alive_sent
      next_ping = @last_packet_received_at + @keep_alive + @ack_timeout if @keep_alive && @keep_alive_sent
      current_time = self.current_time
      [([timeout_from_acks, next_ping].compact.min || current_time) - current_time, 0].max
    end

    def handle_keep_alives
      return unless @keep_alive && @keep_alive > 0

      current_time = self.current_time
      if current_time >= @last_packet_received_at + @keep_alive && !@keep_alive_sent
        packet = MQTT::Packet::Pingreq.new
        send_packet(packet)
        @keep_alive_sent = true
      elsif current_time >= @last_packet_received_at + @keep_alive + @ack_timeout
        raise KeepAliveTimeout
      end
    end

    def puback_packet(packet)
      send_packet(MQTT::Packet::Puback.new(id: packet.id))
    end

    # Read and check a connection acknowledgement packet
    def receive_connack
      Timeout.timeout(@ack_timeout) do
        packet = MQTT::Packet.read(@socket)
        if packet.class != MQTT::Packet::Connack
          raise MQTT::ProtocolException, "Response wasn't a connection acknowledgement: #{packet.class}"
        end

        # Check the return code
        if packet.return_code != 0x00
          # 3.2.2.3 If a server sends a CONNACK packet containing a non-zero
          # return code it MUST then close the Network Connection
          @socket.close
          raise MQTT::ProtocolException, packet.return_msg
        end
        @last_packet_received_at = current_time
        @keep_alive_sent = false
        @connack = packet
      end
    end

    # Send a packet to server
    def send_packet(packet)
      @write_queue << packet
    end

    def parse_uri(uri)
      uri = URI.parse(uri) unless uri.is_a?(URI)
      ssl = case uri.scheme
            when 'mqtt'
              false
            when 'mqtts'
              true
            else
              raise 'Only the mqtt:// and mqtts:// schemes are supported'
            end

      {
        host: uri.host,
        port: uri.port || nil,
        username: uri.user ? URI::Parser.new.unescape(uri.user) : nil,
        password: uri.password ? URI::Parser.new.unescape(uri.password) : nil,
        ssl: ssl
      }
    end

    def next_packet_id
      @last_packet_id = (@last_packet_id || 0).next
      @last_packet_id = 1 if @last_packet_id > 0xffff
      @last_packet_id
    end
  end
end
